# Python Basics
## A Hands-On Tutorial for Movement Neuroscience Graduate Students

---

**Why this tutorial?**  
Before you can analyze EMG signals, build kinematic models, or run statistical tests on reaching data, you need a solid foundation in Python itself. This notebook walks you through the essential building blocks of the language — variables, data types, strings, control flow, and more — so that when you encounter real research code, nothing feels like magic.

Every concept is illustrated with examples relevant to movement neuroscience wherever possible: storing participant data, iterating over trials, handling missing values in experimental recordings, and formatting results for reports.

**Prerequisites:** None — this is the starting point.  
**Environment:** Google Colab (recommended) or Jupyter Notebook.

---

## Table of Contents

1. [Getting Started — The Python Interpreter & Notebooks](#1.-Getting-Started)
2. [Reading Data Files](#2.-Reading-Data-Files)
3. [Tab Completion & Introspection](#3.-Tab-Completion-&-Introspection)
4. [Language Semantics — Indentation & Whitespace](#4.-Language-Semantics)
5. [Variables & Argument Passing](#5.-Variables-&-Argument-Passing)
6. [Mutable vs Immutable Objects](#6.-Mutable-vs-Immutable-Objects)
7. [Scalar Types — Numbers, Strings, Booleans, and None](#7.-Scalar-Types)
8. [Control Flow — if, for, while, and Friends](#8.-Control-Flow)
9. [Exercises](#9.-Exercises)
10. [Summary & Further Reading](#10.-Summary-&-Further-Reading)

---
## 1. Getting Started — The Python Interpreter & Notebooks <a id='1.-Getting-Started'></a>

Python is an **interpreted language**. The Python interpreter runs a program by executing one statement at a time. You can run Python code in several environments:

- **Terminal / Command Line** — type `python` to start the interactive interpreter.
- **IDEs** like Spyder, PyCharm, or VS Code — full-featured editors with debugging, variable explorers, etc.
- **Notebooks** (Google Colab or Jupyter Notebook) — interactive documents that mix code, output, and formatted text. This is what you're using right now.

Google Colab is a hosted Jupyter Notebook service that requires no setup and provides free access to computing resources, including GPUs and TPUs. It's especially well suited for data science and education. In this course, we will use Google Colab, but you are free to use Jupyter Notebooks locally.

Let's start with the absolute basics — assigning a variable and printing it:

In [1]:
# ---- Your first Python statement ----
# Assign the integer 4 to the variable 'a', then print it.

a = 4
print(a)

4


That's it — one line to store a value, one line to display it. Now let's do something slightly more interesting: generate some random numbers using NumPy, a library you'll use constantly in movement neuroscience.

In [2]:
# ---- Printing and importing a library ----
# print() displays output below the cell.
# 'import numpy as np' loads the NumPy library and gives it the shorthand alias 'np'.

print("Hello World")

import numpy as np  # NumPy is the foundation of scientific computing in Python

# Generate 5 random numbers drawn from a standard normal distribution (mean=0, std=1).
# This is a "list comprehension" — a compact way to build a list with a loop.
data = [np.random.standard_normal() for i in range(5)]
print(data)

Hello World
[1.24739231849819, 1.231633122243215, -0.07003575508171664, -0.7899587978852555, 1.7062315934425092]


The first line is a simple `print()` call. The second imports NumPy (which you'll use for everything from storing EMG arrays to computing statistics). The third uses a **list comprehension** — a concise way to generate a list — to draw 5 random samples.

### A note on Colab / Jupyter `.ipynb` files

When you save a Colab notebook (see **File → Save a Copy in Drive**), it creates a file with the `.ipynb` extension. This is a self-contained format that stores all of the content — including code, output, and text — currently in the notebook. Nearly all of the commands and tools in this tutorial work identically in Colab, Jupyter, and traditional IDEs.

---
## 2. Reading Data Files <a id='2.-Reading-Data-Files'></a>

One of the first things you'll need to do in any research project is **load data from a file**. In movement neuroscience, your data might be a `.txt` file of time-series signals, a `.csv` of trial metadata, or a binary file from a motion capture system.

Below is an example of how to mount your Google Drive in Colab and read the first few lines of a text file. If you're running this locally (not in Colab), you can skip the Drive mounting step and just provide a local file path.

**Note:** The code below is specific to Google Colab. If you're using Jupyter locally, replace the file path with a local path on your machine.

In [3]:
# ---- Reading a text file from Google Drive (Colab-specific) ----
# If you are NOT in Colab, skip this cell or adapt the file_path.

import pandas as pd   # Pandas is for tabular data (like spreadsheets)
import numpy as np
import os              # OS module for interacting with the file system

# --- Mount Google Drive (Colab only) ---
# This gives Colab access to files stored in your Google Drive.
# A pop-up will ask you to authorize access.
from google.colab import drive
drive.mount('/content/drive')

# --- Change the working directory to your data folder ---
# Adjust this path to match your own Google Drive folder structure.
os.chdir('/content/drive/MyDrive/Colab Notebooks KINES 597/data/')
!pwd  # Print the current working directory to confirm

# --- Read and display the first 5 lines of a text file ---
file_path = '/content/drive/MyDrive/Colab Notebooks KINES 597/data/sinusoidal_data.txt'

with open(file_path, 'r') as file:
    first_five_lines = file.readlines()[:5]  # Read only the first 5 lines

for line in first_five_lines:
    print(line.strip())  # .strip() removes trailing newline characters

Mounted at /content/drive
/content/drive/MyDrive/Colab Notebooks KINES 597/data
Time(s)	Signal
0.000000	0.000000
0.001000	0.037697
0.002000	0.075380
0.003000	0.113037


### What's happening here?

- **`drive.mount()`** connects your Google Drive so you can access files as if they were local.
- **`os.chdir()`** changes the working directory — like `cd` in a terminal.
- **`with open(file_path, 'r') as file:`** opens the file in read mode (`'r'`). The `with` statement ensures the file is properly closed when you're done, even if an error occurs.
- **`file.readlines()[:5]`** reads all lines into a list, then slices the first 5.
- **`.strip()`** removes whitespace and newline characters from each line.

In practice, for tabular data you'll more often use `pandas.read_csv()` or `numpy.loadtxt()` — but understanding how to read raw text files is a valuable foundation.

---
## 3. Tab Completion & Introspection <a id='3.-Tab-Completion-&-Introspection'></a>

### Tab Completion

Tab completion is a major productivity tool. While typing in a code cell (or in an IDE like Spyder/VS Code), pressing the **Tab** key will search for matching variable names, functions, and methods.

For example, in an IDE or IPython shell:
```python
an_apple = 27
an_example = 42
an  # Press Tab here → shows: an_apple, an_example, any
```

You can also tab-complete **methods on an object**. After typing a variable name followed by a dot (`.`), pressing Tab shows all available methods:
```python
b = [1, 2, 3]
b.  # Press Tab here → shows: append(), count(), insert(), reverse(), sort(), ...
```

Tab completion works slightly differently in notebooks (often showing a dropdown), but the idea is the same.

In [4]:
# ---- Try tab completion yourself ----
# Create a list, then type 'b.' and press Tab to see available methods.

b = [1, 2, 3]

# Referencing a method without calling it (no parentheses) shows its signature:
b.index  # This displays info about the .index() method — it doesn't call it

<function list.index(value, start=0, stop=9223372036854775807, /)>

### Introspection with `?`

Using a **question mark** (`?`) before or after a variable or function displays general information about the object. This works well in IPython and IDEs; in notebooks, the output may appear in a pop-up panel at the bottom.

In [5]:
# ---- Introspection: getting help on objects and functions ----

def add_numbers(a, b):
    """
    Add two numbers together.

    Returns
    -------
    the_sum : type of arguments
    """
    return a + b

# The '?' operator shows the docstring and type information.
# Try running these — output appears in a panel at the bottom of the notebook.
b?
add_numbers?

**Tip for movement neuroscience:** When you're exploring a new library like MNE-Python or OpenSim, tab completion and `?` are your best friends for discovering what methods are available on data objects.

---
## 4. Language Semantics — Indentation & Whitespace <a id='4.-Language-Semantics'></a>

Unlike many other languages (C++, Java, R) that use **braces `{}`** to define code blocks, Python uses **whitespace (indentation)**. A colon (`:`) marks the start of an indented block, and all lines within that block must be indented by the same amount (typically 4 spaces).

This makes Python code visually clean, but you must be careful with indentation — it's not just style, it's **syntax**.

Let's see this in action with a sorting algorithm where indentation defines which statements belong to which loop:

In [6]:
# ---- Demonstrating indentation with Bubble Sort ----
# Notice how indentation level determines which block each line belongs to.

numbers = [5, 2, 9, 1, 5, 6]
print("Unsorted list:", numbers)

# Bubble Sort Algorithm
for i in range(len(numbers)):                          # OUTER loop (level 1)
    for j in range(0, len(numbers) - i - 1):           # INNER loop (level 2)

        if numbers[j] > numbers[j + 1]:                # Condition (level 3)
            # Swap adjacent elements if they're in the wrong order
            numbers[j], numbers[j + 1] = numbers[j + 1], numbers[j]

            # This print is inside the 'if' block (level 4)
            print(f"Swapped {numbers[j + 1]} and {numbers[j]} → {numbers}")

    # This print is inside the OUTER loop but outside the INNER loop (level 1)
    print(f"After pass {i + 1}: {numbers}")

# This print is OUTSIDE all loops (level 0)
print("\nSorted list:", numbers)

Unsorted list: [5, 2, 9, 1, 5, 6]
Swapped 5 and 2 → [2, 5, 9, 1, 5, 6]
Swapped 9 and 1 → [2, 5, 1, 9, 5, 6]
Swapped 9 and 5 → [2, 5, 1, 5, 9, 6]
Swapped 9 and 6 → [2, 5, 1, 5, 6, 9]
After pass 1: [2, 5, 1, 5, 6, 9]
Swapped 5 and 1 → [2, 1, 5, 5, 6, 9]
After pass 2: [2, 1, 5, 5, 6, 9]
Swapped 2 and 1 → [1, 2, 5, 5, 6, 9]
After pass 3: [1, 2, 5, 5, 6, 9]
After pass 4: [1, 2, 5, 5, 6, 9]
After pass 5: [1, 2, 5, 5, 6, 9]
After pass 6: [1, 2, 5, 5, 6, 9]

Sorted list: [1, 2, 5, 5, 6, 9]


### Key rules for indentation

- A **colon** (`:`) denotes the start of an indented code block (after `if`, `for`, `while`, `def`, `class`, etc.).
- Everything in the block must be indented by the **same amount** (4 spaces is standard).
- When the indentation level decreases, you've exited the block.
- **Mixing tabs and spaces** will cause errors — stick to spaces (most editors handle this automatically).

In [7]:
# ---- Multiple statements on one line (semicolons) ----
# You CAN separate statements with semicolons, but this is generally discouraged
# because it reduces readability.

a = 1; b = 2; c = 4  # Legal, but prefer separate lines in real code
print(a, b, c)

1 2 4


---
## 5. Variables & Argument Passing <a id='5.-Variables-&-Argument-Passing'></a>

### Variables are references, not boxes

When you assign a variable in Python, you create a **reference** (a pointer) to the object on the right side of the equals sign. You are **not** copying the data — you're giving it a name.

This distinction is critical when working with large datasets in movement neuroscience. If you accidentally modify a shared reference, you could corrupt your original data.

In [8]:
# ---- Variables are references, not copies ----

a = [11, 23, 34]  # Create a list and assign it to 'a'

In [9]:
# ---- Assigning b = a does NOT copy the list ----
# Both 'a' and 'b' now point to the SAME list object in memory.

b = a
print(b)  # [11, 23, 34]

[11, 23, 34]


In [10]:
# ---- Proof: modifying 'a' also changes 'b' ----
# Because they reference the same object!

a.append(4)  # Add 4 to the list via 'a'
print(b)     # [11, 23, 34, 4] — 'b' sees the change too!

[11, 23, 34, 4]


**This is a common source of bugs.** If you want an independent copy of a list, use `b = a.copy()` or `b = list(a)`. For NumPy arrays, use `b = a.copy()`.

### Everything is an object

An important characteristic of Python is the consistency of its **object model**. Every number, string, data structure, function, class, and module exists as a Python object. Each object has an associated **type** (e.g., `int`, `str`, `list`, `function`) and internal data. This makes the language very flexible — even functions can be stored in variables, passed as arguments, and returned from other functions.

### Passing arguments to functions

Understanding Python's reference semantics is critical when working with large datasets. When you pass an object to a function, the function receives a **reference** to that object, not a copy. This means that if the function modifies the contents of a mutable argument (like a list), the changes are visible outside the function.

In [11]:
# ---- Functions receive references, not copies ----
# This means a function can modify the original list you pass in.

def append_element(some_list, element):
    """Append an element to the given list (modifies it in place)."""
    some_list.append(element)

data = [1, 2, 3]
append_element(data, 4)  # The function modifies 'data' directly
print(data)  # [1, 2, 3, 4] — the original list was changed!

[1, 2, 3, 4]


**Lab analogy:** Imagine `data` is a lab notebook. If you hand someone the actual notebook (a reference), any notes they write go into your notebook. If you hand them a photocopy (a copy), your original stays untouched.

---
## 6. Mutable vs Immutable Objects <a id='6.-Mutable-vs-Immutable-Objects'></a>

Python objects fall into two categories:

| Mutable (can be changed) | Immutable (cannot be changed) |
|---|---|
| `list`, `dict`, `set`, NumPy arrays, most user-defined classes | `int`, `float`, `str`, `tuple`, `frozenset` |

Understanding this distinction helps you avoid unexpected bugs and write safer code.

In [12]:
# ---- Mutable example: lists can be modified in place ----

a_list = ["foo", 2, [4, 5]]
print(f"Before: {a_list}")

a_list[2] = (3, 4)  # Replace the third element
print(f"After:  {a_list}")

Before: ['foo', 2, [4, 5]]
After:  ['foo', 2, (3, 4)]


In [13]:
# ---- Immutable example: tuples CANNOT be modified ----
# This will raise a TypeError — uncomment to see the error.

a_tuple = (3, 5, (4, 5))
# a_tuple[1] = "four"  # TypeError: 'tuple' object does not support item assignment

In [14]:
# ---- Immutable example: strings CANNOT be modified in place ----
# You must create a NEW string instead.

a = "this is a string"
# a[10] = "f"  # TypeError: 'str' object does not support item assignment

# To "modify" a string, use methods that return a NEW string:
b = a.replace("string", "longer string")
print(f"b = {b}")  # New string
print(f"a = {a}")  # Original is unchanged

b = this is a longer string
a = this is a string


**Why this matters in research:** When you load a raw EMG signal into a NumPy array (mutable), any filtering or processing you do modifies the original unless you explicitly make a copy first. A common pattern is:
```python
raw_emg = load_emg_data()       # Original data
filtered_emg = raw_emg.copy()   # Work on a copy
apply_bandpass_filter(filtered_emg)
```

---
## 7. Scalar Types — Numbers, Strings, Booleans, and None <a id='7.-Scalar-Types'></a>

Python has a small set of built-in types for handling numerical data, strings, Boolean (`True`/`False`) values, and the special `None` value. These "single value" types are sometimes called **scalar types**.

### 7.1 Numeric Types: `int` and `float`

In [15]:
# ---- Integers (int) ----
# Python integers can be arbitrarily large — no overflow!

ival = 17239871
print(f"ival ** 7 = {ival ** 7}")  # ** is the exponentiation operator

ival ** 7 = 452624525745445400804783572079381973282879467788991


In [16]:
# ---- Floating-point numbers (float) ----
# Under the hood, each float is a double-precision (64-bit) value.
# You can use scientific notation with 'e'.

fval = 7.243
fval2 = 6.78e-5  # = 0.0000678 (scientific notation)

print(f"fval = {fval}, fval2 = {fval2}")

fval = 7.243, fval2 = 6.78e-05


In [17]:
# ---- Division behavior ----

# Regular division always returns a float:
print(f"5 / 6 = {5 / 6}")    # 0.8333...

# Floor division (//) drops the fractional part:
print(f"7 // 3 = {7 // 3}")  # 2 (integer result)

5 / 6 = 0.8333333333333334
7 // 3 = 2


### 7.2 Strings

Strings are sequences of characters. You'll use them constantly for labeling trials, formatting output, building file paths, and more.

In [18]:
# ---- Creating strings ----
# You can use single quotes or double quotes (double quotes are preferred by convention).

a = 'one way of writing a string'
b = "another way"
print(a)

one way of writing a string


In [19]:
# ---- Multiline strings use triple quotes ----

c = """
This is a longer string that
spans multiple lines
"""
print(c)

# Count the newline characters — there are 3 (before, between, and after the text lines)
print(f"Number of newlines: {c.count(chr(10))}")


This is a longer string that
spans multiple lines

Number of newlines: 3


In [20]:
# ---- Strings are IMMUTABLE ----
# You cannot modify a character in place — you must create a new string.

a = "this is a string"
# a[10] = "f"  # This would raise: TypeError: 'str' object does not support item assignment

# Instead, use .replace() which returns a NEW string:
b = a.replace("string", "longer string")
print(f"b: {b}")
print(f"a: {a}")  # Original is unchanged

b: this is a longer string
a: this is a string


In [21]:
# ---- Converting objects to strings ----
# The str() function converts almost anything to a string.

a = 5.6
s = str(a)
print(f"Value: {s}, Type: {type(s)}")  # '5.6' as a string

Value: 5.6, Type: <class 'str'>


In [22]:
# ---- Strings as sequences: indexing and slicing ----
# Strings behave like lists of characters.

s = "python"
print(f"Characters: {list(s)}")  # ['p', 'y', 't', 'h', 'o', 'n']
print(f"First 3:   {s[:3]}")     # 'pyt' — slicing syntax: s[start:stop]

Characters: ['p', 'y', 't', 'h', 'o', 'n']
First 3:   pyt


In [23]:
# ---- String concatenation ----
# The + operator joins strings together.

a = "this is the first half "
b = "and this is the second half"
print(a + b)

this is the first half and this is the second half


### String Formatting

Python offers two main ways to create formatted strings. The newer **f-string** syntax (introduced in Python 3.6) is generally preferred for readability.

In [24]:
# ---- Method 1: .format() (older style) ----
# {0:.2f} → format first argument as float with 2 decimal places
# {1:s}   → format second argument as string
# {2:d}   → format third argument as integer

template = "{0:.2f} {1:s} are worth US${2:d}"
result = template.format(88.46, "Argentine Pesos", 1)
print(result)

88.46 Argentine Pesos are worth US$1


In [25]:
# ---- Method 2: f-strings (preferred — more readable) ----
# Prefix the string with 'f' and embed expressions directly in {curly braces}.

amount = 10
rate = 88.46
currency = "Pesos"

# You can do math inside the braces and add format specifiers after a colon:
print(f"{amount} {currency} is worth US${amount / rate:.2f}")

# Movement neuroscience example:
participant = "P01"
mean_rt = 0.3127
print(f"Participant {participant}: mean reaction time = {mean_rt:.3f} s")

10 Pesos is worth US$0.11
Participant P01: mean reaction time = 0.313 s


### Common String Methods

Strings have many built-in methods that you'll use constantly in research code — parsing file paths, building condition labels, cleaning data exported from sensors. Here are the most important ones:

| Method | What it does | Example |
|---|---|---|
| `.split(sep)` | Split string into a list at each `sep` | `"a,b,c".split(",")` → `['a','b','c']` |
| `.join(list)` | Join a list into one string with separator | `"_".join(['P01','ctrl'])` → `'P01_ctrl'` |
| `.strip()` | Remove leading/trailing whitespace | `"  hello  ".strip()` → `'hello'` |
| `.replace(old, new)` | Replace all occurrences | `"foo bar".replace("bar","baz")` → `'foo baz'` |
| `.upper()` / `.lower()` | Change case | `"Hello".lower()` → `'hello'` |
| `.startswith(prefix)` | Check if string starts with prefix | `"P01_ctrl".startswith("P01")` → `True` |
| `.endswith(suffix)` | Check if string ends with suffix | `"data.csv".endswith(".csv")` → `True` |

In [ ]:
# ---- .split(): break a string into a list ----
# Extremely common for parsing file paths and CSV-like strings.

path = "/content/drive/MyDrive/data/P01_session2.csv"
parts = path.split("/")           # Split at every '/'
print(f"Path parts: {parts}")
print(f"Filename:   {parts[-1]}")   # Last element = filename

# Split a condition label:
label = "P01_control_trial05"
fields = label.split("_")
print(f"\nFields: {fields}")        # ['P01', 'control', 'trial05']

In [ ]:
# ---- .join(): combine a list into one string ----
# The opposite of .split(). The separator goes BEFORE .join().

parts = ["P01", "perturbation", "block3"]
filename = "_".join(parts) + ".csv"     # 'P01_perturbation_block3.csv'
print(f"Built filename: {filename}")

# Join with a different separator:
print(f"Comma-separated: {', '.join(parts)}")
print(f"No separator:    {''.join(['a', 'b', 'c'])}")

In [ ]:
# ---- .strip(), .startswith(), .endswith() ----
# Useful for cleaning data and filtering files.

# Strip whitespace (common in sensor exports with trailing spaces/newlines):
messy = "  0.3421  \n"
clean = messy.strip()
print(f"Stripped: '{clean}'")

# Check file extensions:
filename = "P01_session2.csv"
print(f"Is CSV?  {filename.endswith('.csv')}")
print(f"Is P01?  {filename.startswith('P01')}")

# Practical: filter a list of filenames
all_files = ["P01_data.csv", "P02_data.csv", "notes.txt", "P01_backup.csv"]
csv_files = [f for f in all_files if f.endswith(".csv")]
p01_files = [f for f in all_files if f.startswith("P01")]
print(f"\nCSV files: {csv_files}")
print(f"P01 files: {p01_files}")

### 7.3 Booleans

The two Boolean values in Python are `True` and `False`. Comparisons and conditional expressions evaluate to one of these. Boolean values are combined with the `and`, `or`, and `not` keywords.

In [26]:
# ---- Boolean logic ----

print(f"True and True  = {True and True}")    # True
print(f"True or False  = {True or False}")     # True
print(f"not True       = {not True}")           # False

# When converted to numbers: False → 0, True → 1
print(f"int(False)     = {int(False)}")         # 0
print(f"int(True)      = {int(True)}")          # 1

# This is useful! e.g., count how many trials were "successful":
trial_results = [True, True, False, True, False]
print(f"Successful trials: {sum(trial_results)} out of {len(trial_results)}")

True and True  = True
True or False  = True
not True       = False
int(False)     = 0
int(True)      = 1
Successful trials: 3 out of 5


### 7.4 Type Casting

The `str`, `bool`, `int`, and `float` types double as **functions** that convert values between types.

In [27]:
# ---- Type casting examples ----

s = "3.14159"              # A string
fval = float(s)            # Convert string → float
print(f"float('{s}') = {fval}, type = {type(fval)}")

ival = int(fval)           # Convert float → int (truncates decimal)
print(f"int({fval}) = {ival}")

bval = bool(fval)          # Any non-zero number → True; 0 → False
print(f"bool({fval}) = {bval}")
print(f"bool(0) = {bool(0)}")

float('3.14159') = 3.14159, type = <class 'float'>
int(3.14159) = 3
bool(3.14159) = True
bool(0) = False


### 7.5 None — Python's "nothing" value

`None` is Python's null value — it represents the absence of a value. It's commonly used as a default for optional function arguments and to indicate missing data.

In [28]:
# ---- None basics ----
# Use 'is' (not ==) to check for None.

a = None
print(f"a is None: {a is None}")       # True
print(f"a is not None: {a is not None}") # False

a is None: True
a is not None: False


In [29]:
# ---- None as a default function argument ----
# This is a very common pattern in Python.
# Movement neuroscience example: optionally scale a force value.

def add_and_maybe_multiply(a, b, c=None):
    """
    Add a + b, and optionally multiply by c if provided.

    Parameters
    ----------
    a, b : numbers to add
    c : optional multiplier (default: None, meaning no multiplication)
    """
    result = a + b
    if c is not None:
        result = result * c
    return result

print(f"add_and_maybe_multiply(2, 3)    = {add_and_maybe_multiply(2, 3)}")      # 5
print(f"add_and_maybe_multiply(2, 3, 4) = {add_and_maybe_multiply(2, 3, 4)}")  # 20

add_and_maybe_multiply(2, 3)    = 5
add_and_maybe_multiply(2, 3, 4) = 20


---
## 8. Control Flow — `if`, `for`, `while`, and Friends <a id='8.-Control-Flow'></a>

Control flow statements let your program make decisions and repeat actions. These are the workhorses of any data analysis pipeline.

### 8.1 `if`, `elif`, `else`

The `if` statement checks a condition. If `True`, it executes the indented block that follows. You can chain additional conditions with `elif` ("else if") and catch everything else with `else`.

In [30]:
# ---- Basic if statement ----

x = -5
if x < 0:
    print("It's negative")  # Note the indent — this line is inside the 'if' block

It's negative


In [31]:
# ---- if / elif / else chain ----
# Only the FIRST matching condition is executed.

x = -5

if x < 0:
    print("It's negative")
elif x == 0:
    print("Equal to zero")
elif 0 < x < 5:  # Python allows chained comparisons!
    print("Positive but smaller than 5")
else:
    print("Positive and larger than or equal to 5")

It's negative


In [32]:
# ---- Chained comparisons ----
# Python lets you write mathematical-style chains like 1 < x < 10.

print(f"3 > 1 > 5 is {3 > 1 > 5}")  # False (because 1 > 5 is False)
print(f"1 < 3 < 5 is {1 < 3 < 5}")  # True

3 > 1 > 5 is False
1 < 3 < 5 is True


### 8.2 `for` Loops

`for` loops iterate over a collection (like a list, tuple, or range). They're the primary way to process data trial-by-trial, sample-by-sample, or participant-by-participant.

In [33]:
# ---- for loop with 'continue' ----
# 'continue' skips the rest of the current iteration and moves to the next one.
# This is useful for skipping missing data (None values).

# Imagine these are force readings, with some missing values:
sequence = [1, 2, None, 4, None, 5]
total = 0

for value in sequence:
    if value is None:
        continue       # Skip this iteration — go to the next value
    total += value     # Only runs if value is not None

print(f"Sum (skipping None): {total}")  # 12

Sum (skipping None): 12


In [34]:
# ---- for loop with 'break' ----
# 'break' exits the loop entirely.
# Example: sum values until you hit a 5 (a sentinel value).

sequence = [1, 2, 0, 4, 6, 5, 2, 1]
total_until_5 = 0

for value in sequence:
    if value == 5:
        break                # Stop the loop immediately
    total_until_5 += value

print(f"Sum until 5: {total_until_5}")  # 1+2+0+4+6 = 13

Sum until 5: 13


In [35]:
# ---- break only exits the INNERMOST loop ----
# Outer loops continue running.

for i in range(4):
    for j in range(4):
        if j > i:
            break          # Exits the inner loop only
        print((i, j))

(0, 0)
(1, 0)
(1, 1)
(2, 0)
(2, 1)
(2, 2)
(3, 0)
(3, 1)
(3, 2)
(3, 3)


### 8.3 `while` Loops

A `while` loop repeats a block of code as long as its condition remains `True`. Use `break` to exit early.

In [36]:
# ---- while loop with a break condition ----
# Keep halving x and accumulating the total, but stop if total exceeds 500.

x = 256
total = 0

while x > 0:
    if total > 500:
        break          # Safety valve — stop if total gets too large
    total += x
    x = x // 2         # Floor division: halve x each iteration
    print(f"x = {x}, total = {total}")

x = 128, total = 256
x = 64, total = 384
x = 32, total = 448
x = 16, total = 480
x = 8, total = 496
x = 4, total = 504


### 8.4 `pass` — The "Do Nothing" Statement

`pass` is a placeholder that does nothing. You use it when Python requires a statement (e.g., inside an `if` or `class` block) but you haven't written the code yet.

In [37]:
# ---- pass as a placeholder ----

x = 128  # From the while loop above

if x < 0:
    print("negative!")
elif x == 0:
    # TODO: handle the zero case later
    pass              # Placeholder — without this, Python would raise a SyntaxError
else:
    print("positive!")

positive!


### 8.5 `range()` — Generating Sequences of Integers

The `range()` function generates a sequence of evenly spaced integers. It's the most common way to create loop counters.

In [38]:
# ---- range() basics ----

# range(n) generates integers from 0 to n-1
print(f"range(10): {list(range(10))}")        # [0, 1, 2, ..., 9]

# range(start, stop, step)
print(f"range(0, 20, 2): {list(range(0, 20, 2))}")  # Even numbers: [0, 2, 4, ..., 18]
print(f"range(5, 0, -1): {list(range(5, 0, -1))}")  # Countdown: [5, 4, 3, 2, 1]

range(10): [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
range(0, 20, 2): [0, 2, 4, 6, 8, 10, 12, 14, 16, 18]
range(5, 0, -1): [5, 4, 3, 2, 1]


In [39]:
# ---- Using range() with a for loop ----
# A common pattern for iterating with an index.

seq = [1, 2, 3, 4]
for i in range(len(seq)):
    print(f"element {i}: {seq[i]}")

# Note: In practice, enumerate() is often preferred (covered in later tutorials):
# for i, val in enumerate(seq):
#     print(f"element {i}: {val}")

element 0: 1
element 1: 2
element 2: 3
element 3: 4


---
## 9. Exercises <a id='9.-Exercises'></a>

Test your understanding by completing these exercises. Each one uses the concepts from this tutorial.

---

### Exercise 1: Variable References

Predict the output of the following code **before** running it. Then run it to check.

```python
a = [10, 20, 30]
b = a
b.append(40)
a[0] = 999
print(a)
print(b)
```

Then modify the code so that `b` is an **independent copy** of `a` (changing `b` should not affect `a`).

In [40]:
# ---- YOUR CODE HERE ----

# Step 1: Run the original code and observe
a = [10, 20, 30]
b = a
b.append(40)
a[0] = 999
print(f"a = {a}")
print(f"b = {b}")

# Step 2: Now fix it so b is independent of a.
# TODO: Replace 'b = a' with something that makes an independent copy.
# Hint: use .copy() or list()

a = [999, 20, 30, 40]
b = [999, 20, 30, 40]


### Exercise 2: Filtering Trial Data

You have a list of reaction times (in seconds) from a reaching experiment. Some entries are `None` (failed trials). Write a `for` loop that:
1. Skips `None` values using `continue`.
2. Computes the **mean** of the valid reaction times.
3. Prints the result formatted to 3 decimal places.

In [ ]:
# ---- YOUR CODE HERE ----

reaction_times = [0.312, 0.298, None, 0.345, 0.289, None, 0.331, 0.305]

total = 0
count = 0

for rt in reaction_times:
    # TODO: skip None values
    # TODO: accumulate total and count
    pass

# TODO: compute mean and print
# Expected output: something like "Mean RT: 0.313 s (6 valid trials)"

### Exercise 3: String Formatting for a Lab Report

Given the variables below, use an **f-string** to produce the following output:

```
Participant P03 (age 24, right-handed)
  Mean endpoint error: 2.47 cm
  Trials completed: 48 / 50
```

In [ ]:
# ---- YOUR CODE HERE ----

participant_id = "P03"
age = 24
handedness = "right"
mean_error = 2.4723
trials_completed = 48
total_trials = 50

# TODO: Write f-string print statements to produce the formatted output above.
# Hint: use :.2f for the error to get 2 decimal places

### Exercise 4: Classifying Trial Outcomes (Challenge)

Write a `for` loop that iterates over the list of endpoint errors below. For each trial, classify it as:
- `"excellent"` if error < 1.0 cm
- `"good"` if 1.0 ≤ error < 2.5 cm
- `"poor"` if error ≥ 2.5 cm

Print the trial number and classification for each trial, then print the count of each category.

In [ ]:
# ---- YOUR CODE HERE ----

endpoint_errors = [0.8, 1.3, 2.7, 0.5, 1.9, 3.1, 0.9, 2.4, 1.1, 0.3]

excellent_count = 0
good_count = 0
poor_count = 0

for i in range(len(endpoint_errors)):
    error = endpoint_errors[i]
    # TODO: classify the error and print the result
    # TODO: increment the appropriate counter
    pass

# TODO: print the summary counts

---
## 10. Summary & Further Reading <a id='10.-Summary-&-Further-Reading'></a>

### What You Learned

| Concept | What it does | Example |
|---|---|---|
| **Variables** | Store references to objects | `x = 42`, `name = "Alice"` |
| **`print()`** | Display output | `print(f"RT = {rt:.3f} s")` |
| **`import`** | Load external libraries | `import numpy as np` |
| **Indentation** | Defines code blocks (not braces!) | 4 spaces per level |
| **References** | `b = a` makes `b` point to same object | Use `.copy()` for independence |
| **Mutable** | Can be changed in place | `list`, `dict`, NumPy arrays |
| **Immutable** | Cannot be changed in place | `int`, `float`, `str`, `tuple` |
| **`int`, `float`** | Numeric types | `5`, `3.14`, `6.78e-5` |
| **`str`** | Text strings | `"hello"`, `f"RT = {val:.2f}"` |
| **String methods** | `.split()`, `.join()`, `.strip()`, `.startswith()` | Parse paths, build labels, clean data |
| **`bool`** | True / False | `True and False` → `False` |
| **`None`** | Absence of a value | Default arguments, missing data |
| **`if/elif/else`** | Conditional branching | Check trial conditions |
| **`for` loop** | Iterate over collections | Loop through trials |
| **`while` loop** | Repeat while condition is true | Adaptive algorithms |
| **`break`** | Exit a loop early | Stop at a sentinel value |
| **`continue`** | Skip to next iteration | Skip missing data (None) |
| **`pass`** | Do nothing (placeholder) | Empty code blocks |
| **`range()`** | Generate integer sequences | `range(10)`, `range(0, 100, 5)` |

### Further Reading

- [Python Official Tutorial](https://docs.python.org/3/tutorial/) — the authoritative beginner guide
- [Real Python](https://realpython.com/) — excellent, in-depth tutorials on every Python topic
- [NumPy Quickstart](https://numpy.org/doc/stable/user/quickstart.html) — essential for scientific computing
- [Pandas Getting Started](https://pandas.pydata.org/docs/getting_started/intro_tutorials/) — for tabular data (trials, participants, metadata)
- [Google Colab Introduction](https://colab.research.google.com/notebooks/intro.ipynb) — getting the most out of Colab

### What's Next?

With these fundamentals in place, you're ready to tackle:
- **Data structures** — lists, dictionaries, tuples, sets
- **Functions** — writing reusable analysis code
- **NumPy & Pandas** — the backbone of scientific data analysis
- **Object-Oriented Programming** — modeling experiments, participants, and signals as classes

---

*Happy coding, and may your p-values be small and your effect sizes large!*